### PR TITLE
add authentication server

### PR DESCRIPTION
**Pull request by andrei-astaschenko**

### CODE REVIEW

File: auth-server/build.gradle [hunk 1]
```diff
@@ -0,0 +1,21 @@
+version = '0.0.1-SNAPSHOT'
```
Comment (Andrei-Khadziukou) [inline auth-server/build.gradle]: Product file names should follow the mask: <artefact_name>-<build_number>.extention

It looks like we should use '0.0.1.SNAPSHOT', @susau is it correct?

File: auth-server/src/main/resources/application.yml [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+    cache: false
+
+server:
+  context-path: /uaa
+  port: 8085
+
+eureka:
+  client:
+    serviceUrl:
+      defaultZone: http://127.0.0.1:8080/eureka/
```
Comment (Andrei-Khadziukou) [inline auth-server/src/main/resources/application.yml]: lets use the environment variable with default value to have the ability to customize this property

File: gateway/src/main/java/com/epam/petclinic/zuul/GatewayApplication.java [hunk 3]
```diff
@@ -2,6 +2,7 @@
 
 import org.springframework.boot.SpringApplication;
 import org.springframework.boot.autoconfigure.SpringBootApplication;
+import org.springframework.boot.autoconfigure.security.oauth2.client.EnableOAuth2Sso;
```
Comment (Andrei-Khadziukou) [inline gateway/src/main/java/com/epam/petclinic/zuul/GatewayApplication.java]: I see import but there is no usage

File: auth-server/src/main/resources/application.yml [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+  application:
+     name: AUTH-SERVICE
+  freemarker:
+    enabled: false
+  thymeleaf:
+    cache: false
+
+server:
+  context-path: /uaa
+  port: 8085
```
Comment (Andrei-Khadziukou) [inline auth-server/src/main/resources/application.yml:14]: Could you please add the 'Allocated ports' table to the our POC document and fill the first record something like AUTH-SERVER | 8085

File: auth-server/src/main/resources/application.yml [hunk 5]
```diff
@@ -0,0 +1,21 @@
+info:
+  component: Authentication Service
+
+spring:
+  application:
+     name: AUTH-SERVICE
```
Comment (Andrei-Khadziukou) [inline auth-server/src/main/resources/application.yml]: From my point of view the 'AUTH-SERVER' name will be better

File: auth-server/src/main/java/com/epam/petclinic/authentication/WebSecurityConfig.java [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+import org.springframework.security.authentication.AuthenticationManager;
+import org.springframework.security.config.annotation.authentication.builders.AuthenticationManagerBuilder;
+import org.springframework.security.config.annotation.web.configuration.WebSecurityConfigurerAdapter;
+
+@Configuration
+class WebSecurityConfig extends WebSecurityConfigurerAdapter {
+
+    @Override
+    @Bean
+    public AuthenticationManager authenticationManagerBean() throws Exception {
```
Comment (Andrei-Khadziukou) [inline auth-server/src/main/java/com/epam/petclinic/authentication/WebSecurityConfig.java:14]: Do we really need to override it?

File: auth-server/src/main/java/com/epam/petclinic/authentication/WebSecurityConfig.java [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+        return super.authenticationManagerBean();
+    }
+
+    @Override
+    protected void configure(AuthenticationManagerBuilder auth) throws Exception {
+        auth.inMemoryAuthentication()
```
Comment (Andrei-Khadziukou) [inline auth-server/src/main/java/com/epam/petclinic/authentication/WebSecurityConfig.java]: Add the comment what happens here

Comment (Andrei-Khadziukou) [reply auth-server/src/main/java/com/epam/petclinic/authentication/WebSecurityConfig.java]: Add `//TODO find the way to pass encrypted values as passwords`

File: auth-server/src/main/java/com/epam/petclinic/authentication/AuthorizationServerConfig.java [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+    }
+
+    @Bean
+    public TokenStore tokenStore() {
+        return new JwtTokenStore(jwtTokenEnhancer());
+    }
+
+    @Bean
+    protected JwtAccessTokenConverter jwtTokenEnhancer() {
+        KeyStoreKeyFactory keyStoreKeyFactory = new KeyStoreKeyFactory(new ClassPathResource("jwt.jks"),
```
Comment (Andrei-Khadziukou) [inline auth-server/src/main/java/com/epam/petclinic/authentication/AuthorizationServerConfig.java]: Add a comment with short explanation and `//TODO encrypt 'mySecretKey'`

File: auth-server/src/main/java/com/epam/petclinic/authentication/AuthorizationServerConfig.java [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+    /**
+     * Configures the OAuth2 endpoints to adjust the authentication manager which will represent the web-security users,
+     * and JWT token store configuration.
+     *
+     * @param endpoints @see AuthorizationServerEndpointsConfigurer.
+     * @throws Exception @see Exception on any error.
+     */
+    @Override
+    public void configure(AuthorizationServerEndpointsConfigurer endpoints) throws Exception {
+        endpoints.tokenStore(tokenStore()).tokenEnhancer(jwtTokenEnhancer()).authenticationManager(authenticationManager);
```
Comment (Andrei-Khadziukou) [inline auth-server/src/main/java/com/epam/petclinic/authentication/AuthorizationServerConfig.java]: Different instances of tokenEnhancer for tokenStore and tokenEnhencer of endpoint?

File: auth-server/src/main/java/com/epam/petclinic/authentication/AuthorizationServerConfig.java [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+@EnableAuthorizationServer
+public class AuthorizationServerConfig extends AuthorizationServerConfigurerAdapter {
+
+    @Autowired
+    @Qualifier("authenticationManagerBean")
+    private AuthenticationManager authenticationManager;
+
+    @Override
+    public void configure(ClientDetailsServiceConfigurer clients) throws Exception {
+        clients.inMemory()
```
Comment (Andrei-Khadziukou) [inline auth-server/src/main/java/com/epam/petclinic/authentication/AuthorizationServerConfig.java]: Add comments to make it a bit more clear

Comment (Andrei-Khadziukou) [reply auth-server/src/main/java/com/epam/petclinic/authentication/AuthorizationServerConfig.java]: Add `//TODO encrypt the value for 'secret' `

File: auth-server/src/main/java/com/epam/petclinic/authentication/WebSecurityConfig.java (hunk 6 above)
Comment (andrei-astaschenko) [reply auth-server/src/main/java/com/epam/petclinic/authentication/WebSecurityConfig.java:14]: Yes, We need to do it to create bean and inject object to AuthorizationServerConfig class.

File: auth-server/src/main/resources/application.yml (hunk 4 above)
Comment (andrei-astaschenko) [reply auth-server/src/main/resources/application.yml:14]: Table with service addresses is added.

File: auth-server/src/main/java/com/epam/petclinic/authentication/AuthorizationServerConfig.java [hunk 11]
```diff
@@ -0,0 +1,72 @@
...
+public class AuthorizationServerConfig extends AuthorizationServerConfigurerAdapter {
+
+    @Autowired
+    private AuthenticationManager authenticationManager;
+
+    @Autowired
+    private JwtAccessTokenConverter jwtAccessTokenConverter;
+
+    /*
+        Configures permissions for clients in memory of authentication server and passes the authorities for resource server.
```
Comment (Andrei-Khadziukou) [inline auth-server/src/main/java/com/epam/petclinic/authentication/AuthorizationServerConfig.java]: Could you please add more detail description? What does `scope` mean and so on?

File: auth-server/src/main/resources/application.yml [hunk 12]
```diff
@@ -0,0 +1,21 @@
...
+  application:
+     name: AUTH-SERVER
+  freemarker:
+    enabled: false
+  thymeleaf:
+    cache: false
+
+server:
+  context-path: /uaa
+  port: ${service.auth.server.port:8085}
```
Comment (Andrei-Khadziukou) [inline auth-server/src/main/resources/application.yml]: no needs in property here, can be configured via config server i think

File: gateway/src/main/resources/application.yml [hunk 13]
```diff
@@ -18,7 +18,12 @@ eureka:
 
 zuul:
   prefix: /api
+  sensitiveHeaders:
```
Comment (Andrei-Khadziukou) [inline gateway/src/main/resources/application.yml]: empty?

File: auth-server/src/main/java/com/epam/petclinic/authentication/AuthorizationServerConfig.java [hunk 14]
```diff
@@ -0,0 +1,81 @@
...
+public class AuthorizationServerConfig extends AuthorizationServerConfigurerAdapter {
+
+    @Autowired
+    private AuthenticationManager authenticationManager;
+
+    @Autowired
+    private JwtAccessTokenConverter jwtAccessTokenConverter;
+
+    /**
+     *          Configures permissions for clients in memory of authentication server and passes the authorities for
```
Comment (Andrei-Khadziukou) [inline auth-server/src/main/java/com/epam/petclinic/authentication/AuthorizationServerConfig.java]: Why do you use a such formatting?

File: gateway/src/main/resources/application.yml [hunk 15]
```diff
@@ -18,7 +18,12 @@ eureka:
 
 zuul:
   prefix: /api
+  sensitiveHeaders: Cookie,Set-Cookie
```
Comment (Andrei-Khadziukou) [inline gateway/src/main/resources/application.yml]: Add a comment to clarify why we need it

**Pull request merged**